cli/src/commands/extensions/extension.rs

```diff
@@ -0,0 +1,143 @@
+use std::{convert::TryFrom, fs::File, io::Read, path::PathBuf};
+
+use anyhow::{anyhow, Result};
+use serde::Deserialize;
+
+const MANIFEST_NAME: &str = "PhylumExt.toml";
+
+#[derive(Debug)]
+pub struct Extension {
+    path: PathBuf,
+    manifest: ExtensionManifest,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct ExtensionManifest {
+    name: String,
+    description: Option<String>,
+    entry_point: String,
+}
+
+impl Extension {
+    pub fn name(&self) -> &str {
+        &self.manifest.name
+    }
+
+    pub fn description(&self) -> Option<&str> {
+        self.manifest.description.as_deref()
+    }
+
+    /// Install the extension in the default path.
+    pub fn install(&self) -> Result<()> {
+        println!("Installing extension {}...", self.name());
+        let target_prefix = extensions_path()?.join(self.name());
+
+        if target_prefix == self.path {
+            return Err(anyhow!(
+                "extension path and installation path are identical, skipping"
+            ));
+        }
+
+        for entry in walkdir::WalkDir::new(&self.path) {
+            let source_path = entry?.into_path();
+            let dest_path = target_prefix.join(source_path.strip_prefix(&self.path)?);
+
+            if source_path.is_dir() {
+                std::fs::create_dir_all(dest_path)?;
+            } else if source_path.is_file() {
+                if dest_path.exists() {
+                    return Err(anyhow!("{}: already exists", dest_path.to_string_lossy()));
+                } else {
+                    std::fs::copy(source_path, dest_path)?;
+                }
+            }
+        }
+
+        println!("Extension {} installed correctly", self.name());
+
+        Ok(())
+    }
+
+    pub fn uninstall(self) -> Result<()> {
+        println!("Uninstalling extension {}...", self.name());
+        let target_prefix = extensions_path()?.join(self.name());
+
+        if target_prefix != self.path {
+            println!("{:?} {:?}", target_prefix, self.path);
+            return Err(anyhow!(
+                "extension {} is not installed, skipping",
+                self.name()
+            ));
+        }
+
+        for entry in walkdir::WalkDir::new(&self.path).contents_first(true) {
+            let entry = entry?.into_path();
+            println!("removing {}...", entry.to_string_lossy());
+            if entry.is_dir() {
+                std::fs::remove_dir(entry)?;
+            } else if entry.is_file() {
+                std::fs::remove_file(entry)?;
+            }
+        }
```

Comment:
Should probably just remove the directories including all contents, rather than walking the directory and deleting everything individually?

Reply:
There is `std::fs::remove_dir_all` that would serve this purpose, but which would fail were we to switch to the single-file format (granted, it would be a very minor change and easy to catch with tests). I chose this approach in the interest of being more explicit and intentional, and also having the chance of listing the removed files to the user. Furthermore, this explicitly breaks support for symlinks, which we may not want to have in the first place and whose side effects I'm not confident about. I think switching to `remove_dir_all` would be fine still if we feel like these points aren't that important (or are counterproductive).

Reply:
Those are all solid arguments, but they seem mostly focused on what *might* be needed rather than what *is* needed. I'd recommend just going for the best solution now, rather than the best solution in the future and switch to `remove_dir_all`.

Reply:
That makes a lot of sense; I do have a tendency to cover as much unknowns as I can, but on the other hand it's good to keep the scope tight whenever possible.